code/modules/fish/fish_tank.dm

```diff
@@ -340,19 +355,43 @@
 		fish_list.Remove(pick(fish_list))			//Kill a random fish
 	update_icon()
 
+/obj/machinery/fishtank/proc/seadevil_eat()
+	var/tmp/list/fish_to_eat = fish_list.Copy()
+	fish_to_eat.Remove("sea devil")
+	var/eat_target = pick(fish_to_eat)
+	visible_message("<span class='notice'>The sea devil devours \an [eat_target].</span>")
+	kill_fish(eat_target)
+
 /obj/machinery/fishtank/proc/add_fish(var/type)
+	if(!type || type == "dud")
+		return
 	//Check if we were passed a fish type
 	fish_list.Add("[type]")						//Add a fish of the specified type
 	//Announce the new fish
 	update_icon()
-	visible_message("A new [type] has hatched in \the [src]!")
+	if(nonhatching_types.Find(type))
+		visible_message("The [type] has been placed in \the [src]!")
+	else
+		visible_message("A new [type] has hatched in \the [src]!")
 
 /obj/machinery/fishtank/proc/select_egg_type()
-	var/fish = pick(fish_list)						//Select a fish from the fish in the tank
-	if(prob(25))									//25% chance to be a dud (blank) egg
+	var/fish = null
+	if(prob(10)) //Small chance for infertility
 		fish = "dud"
+	else
+		fish = recursive_valid_egg(fish_list)
 	var/obj/item/fish_eggs/egg_path	= fish_eggs_list[fish]					//Locate the corresponding path from fish_eggs_list that matches the fish
-	return egg_path									//The fish was located in the fish_eggs_list, so return the proper egg
+	return egg_path
+
+/obj/machinery/fishtank/proc/recursive_valid_egg(var/list/pick_egg_from)
+	var/fish = pick(pick_egg_from)
+	if(!fish || nonhatching_types.Find(fish))
```

Comment:
`!fish` can never happen here. You also never check if you have an empty list, which will cause a runtime if passed to `pick()`. Why does this need to be a recursive check again?

Reply:
The reason it was written this way is somewhat complicated, because fish aren't OOP. I wrote it this way so that other people could add seadevil-like nonhatching fish in the future if they wanted to. I suppose a non recursive way to do this would be

```
var/list/pick_egg_from = egg_list.Copy()
for(var/invalid in nonhatching_types)
   pick_egg_from.Remove(invalid)
return pick(pick_egg_from)
```

The !fish check is actually because I kept getting a weird bug where [empty string] eggs would be rarely laid. Very strange behavior but this fixed it.

Reply:
Okay, perhaps it would be better to figure out where those [empty string] eggs are coming from

Reply:
Do you want me to use the non-recursive implementation or does it matter?

Reply:
whatever